### PR TITLE
Leo/cha 509 handle backend error responses

### PR DESCRIPTION
**Pull request by lmangall**
**This first PR handles only 409 for login endpoint**
-> If we merge it the approach will be extended to more codes and more API calls

When there is an error the backend closes the session but the user is still in the app. The front end did not notify the user about the errors and there is unclear information about what is happening.

- [x] add `currentTab` to user context 
- [x] use user context to switch tab
- [x] create a new component to handle backend errors : `BackendError` When an API call is done, the component will be triggered in the case of an error
- [x] Test the component with login endpoint
- [x] support 409 in  `BackendError` and in the mock for easy testing in dev and prod

## Custom error interface
```
interface CustomError extends Error {
  status?: number;
}
```

The `CustomError` interface extends the built-in Error interface in TypeScript to include an optional status property. 

## Catching the error

state declaration uses _useState_ hook. It can hold an object with a _message_ and an _errorCode_
`const [error, setError] = useState<{ message: string; errorCode: number } | null>(null);
`

```
  try {
    const response = await fetch(`${backendUrl}/send-code`, {
.......
    });
    if (!response.ok) {
      const errorMessage = await response.text();
      throw new Error(errorMessage);
    }
```

Catch Block: If an error occurs during the fetch the setError function is called to update the error state with a message and an error code.

## Rendering the error
```
{error && (
  <BackendError
    message={error.message}
    errorCode={error.errorCode}
    onClose={() => setError(null)}
    onRedirect={() => setIsPhoneSubmitted(false)}
  />
)}
```
The BackendError component is rendered if the error state is not null.

### CODE REVIEW

**Comment by 552020** [comment]
Can you explain in simpler terms what is the idea behind your strategy? The /login returns 409, which means that the user is already logged in, so what you want to do in this case? Is this an error at all? I mean if the user is already logged in and the user wants to log in, the endpoint could just return the chats of the user again and everything is fine no?

**Comment by 552020** [comment]
Can you explain in simpler terms what is the idea behind your strategy? The /login returns 409, which means that the user is already logged in, so what you want to do in this case? Is this an error at all? I mean if the user is already logged in and the user wants to log in, the endpoint could just return the chats of the user again and everything is fine no?

**Comment by lmangall** [comment]
There is a misunderstanding in the purpose of this PR, let's just say on this PR I want to "handle" the errors, not "manage" them (the managing of 409 is just a demo)

So the goal is simply to show the message to the user and eventually redirect to home when the session is abandoned by backend (and we haven't even decided yet on how the backend will communicate it,[ see here](https://linear.app/chatpay/issue/CHA-509/handle-backend-error-responses#comment-96e61e33) )

This PR is a test to validate the approach (in the description).
Most of the error are not supposed to happen or not happening so it is a problem to reproduce them, but with 409 it's not a problem to test both in development and in production.
If this approach is validated I will extend the error catching to all API calls in a next PR, and enhance `BackendError` so it at least redirect to home if a session is terminated (again this will have to be decided with backend) but also handles error in different and individual ways.

Coming back to your question about precisely 409, this is still unimplemented by backend this is mostly  https://linear.app/chatpay/issue/CHA-517/implement-user-leave-app-and-come-back-while-session-active
But also another Issue where Daniil save chats in DB during a session -> this is done BUT there is no way yet to retrieve it for front end. At the moment if you connect (fully) then leave the app and come back while your session is on you are completely stuck. @Danielg1406 is familiar with the problem and this is the reason why active session time has been reduced from 15mn to 7mn. 
It's also in link with @Danielg1406  last implementation to store the chats in the context (this will facilitate our user coming back to the app after closing it)

**Comment by 552020** [comment]
> There is a misunderstanding in the purpose of this PR, let's just say on this PR I want to "handle" the errors, not "manage" them (the managing of 409 is just a demo)
> 

What is the difference between 'handle' and 'manage'? 'Handle' an error means something like 'manage' but without understanding what this error is about?

> So the goal is simply to show the message to the user and eventually redirect to home when the session is abandoned by backend (and we haven't even decided yet on how the backend will communicate it,[ see here](https://linear.app/chatpay/issue/CHA-509/handle-backend-error-responses#comment-96e61e33) )
> 

In the case of 409 I don't have the feeling that the session is abandoned by the backend. Anyway I commented also there about this 409 returned by /login. I mean: I didn't pick up myself this. particular error to start with. 

> This PR is a test to validate the approach (in the description). Most of the error are not supposed to happen or not happening so it is a problem to reproduce them, but with 409 it's not a problem to test both in development and in production. If this approach is validated I will extend the error catching to all API calls in a next PR, and enhance `BackendError` so it at least redirect to home if a session is terminated (again this will have to be decided with backend) but also handles error in different and individual ways.
> 

Since the approach is not clear to me, I was asking you to better clarify what is your strategy in the case of an error. Like where variable are changed and why, where is the user brought and so on. 


> Coming back to your question about precisely 409, this is still unimplemented by backend this is mostly https://linear.app/chatpay/issue/CHA-517/implement-user-leave-app-and-come-back-while-session-active But also another Issue where Daniil save chats in DB during a session -> this is done BUT there is no way yet to retrieve it for front end. At the moment if you connect (fully) then leave the app and come back while your session is on you are completely stuck. @Danielg1406 is familiar with the problem and this is the reason why active session time has been reduced from 15mn to 7mn. It's also in link with @Danielg1406 last implementation to store the chats in the context (this will facilitate our user coming back to the app after closing it)

What does it mean that /login 409 is still not implemented yet by the backend? Can you confirm? @dantol29. It's clear to me that at the moment login 409 doesn't give back anything you can work with. But I don't see the reason to start implementing the errors starting from /login 409 where is not clear if the session is closed or not and write some code that will be changed in a couple of hours, if Daniil gives back the chats also in the case of 409.

**Review by 552020 — APPROVED** [review]
LGTM!
